internal/dns/dns.go

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package dns
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/miekg/dns"
+)
+
+const (
+	// Defaults taken from /etc/resolv.conf man page
+	defaultDNSOptionNdots    = 1
+	defaultDNSOptionTimeout  = 5
+	defaultDNSOptionAttempts = 2
+
+	defaultEtcResolvConfFile   = "/etc/resolv.conf"
+	consulDataplaneDNSBindHost = "127.0.0.1"
+)
+
+type ConfigureConsulDNSInput struct {
+	// Used only for unit tests
+	etcResolvConfFile string
+}
+
+// ConfigureConsulDNS reconstructs the /etc/resolv.conf file by setting the
+// consul-dataplane's DNS server (i.e. localhost) as the first nameserver in the list.
+func (i *ConfigureConsulDNSInput) ConfigureConsulDNS() error {
+	etcResolvConfFile := defaultEtcResolvConfFile
+	if i.etcResolvConfFile != "" {
+		etcResolvConfFile = i.etcResolvConfFile
+	}
+
+	cfg, err := dns.ClientConfigFromFile(etcResolvConfFile)
+	if err != nil {
+		return err
+	}
+
+	if cfg == nil {
+		return fmt.Errorf("failed to fetch DNS config")
+	}
+
+	options := constructDNSOpts(cfg)
+
+	nameservers := []string{consulDataplaneDNSBindHost}
+	nameservers = append(nameservers, cfg.Servers...)
```

Comment:
what if nameserver `consulDataplaneDNSBindHost` is already present?

Reply:
It is not expected for localhost to present as a `nameserver`. By default ECS tasks point to a global nameserver for resolving IPs. Hypothetically if we were to encounter such a case, it should just be a duplicate nameserver entry and the lookup should fallback to the subsequent entries in the list. The aim here is to just move the localhost nameserver to the top of the list to make sure we fallback from there on.